contributors/design-proposals/grow-volume-size.md

```diff
@@ -0,0 +1,159 @@
+| Host Path       | No                | No                          | No                       |
+| GlusterFS       | Yes               | No                          | Yes                      |
+| Azure File      | No                | No                          | No                       |
+| Cephfs          | No                | No                          | No                       |
+| NFS             | No                | No                          | No                       |
+
+
+## Implementation Design
+
+For volume type that supports growing the PV size, this will be a two step operation:
+
+* A controller in master-controller will listen for PVC events and perform corresponding cloudprovider operation. If successful - controller will store new device size in PV. Some cloudproviders (such as cinder) - do not allow resizing of attached volumes. In such cases - it is upto volume plugin maintainer to decide appropriate behaviour. Volume Plugin maintainer can choose to ignore resize request if disk is attached to a pod (and add appropriate error events to PVC object).  Resize request will keep failing until user corrects the error. User can take necessary action in such cases (such as scale down the pod) which will allow resize to proceed normally.
+
+  In case where volume type requires no file system resize, both PV & PVC objects will be updated accordingly and `status.capacity` of both objects will reflect new size.
+  For volume plugins that require file system resize - an additional annotation called `volume.alpha.kubernetes.io/fs-resize-pending` will be added to PV to communicate
+  to the Kubelet that File system must be resized when a new pod is started using the PV.
+
+* In case volume plugin doesn’t support resize feature. The resize API request will be rejected and PVC object will not be saved. This check will be performed via an admission controller plugin.
+
+* In case requested size is smaller than current size of PVC. A validation will be used to reject the API request. (This could be moved to admission controller plugin too.)
+
+* There will be additional checks in controller that grows PV size - to ensure that we do not make cloudprovider API calls that can reduce size of PV.
+
+* To consider cases of missed PVC update events, an additional loop will reconcile bound PVCs with PVs.
+
+* Resource Quota code in admission controller has to be updated to consider PVC updates.
+
+* The resize of file system will be performed on kubelet. If there is a running pod - no operation will be performed. Only when a new pod is started using same PVC - then kubelet will match device size and size of pv and attempt a resize of file system.  resizing filesystem will be a volume plugin function. It is upto volume plugin maintainer to correctly implement this. In following cases no resize will be necessary and hence volume plugin can return success without actually doing anything.
```

Comment:
Please also consider how resize would work for PVCs exposed as raw block devices.  While we don't have this feature today it is also being designed and discussed as part of the [local storage proposal](https://github.com/kubernetes/community/pull/306).  In the case of raw devices we can also allow online resize since the pod itself would be responsible for rescanning the device.